src/buildingpart.js

```diff
@@ -407,60 +407,22 @@ class BuildingPart {
   }
 
   /**
-   * Convert a cardinal direction (ESE) to degrees 112°.
-   * North is zero.
+   * Convert a cardinal direction to degrees.
+   * North is zero and values increase clockwise.
+   *
+   * @param {string} cardinal - the direction.
+   *
+   * @return {int} degrees
    */
   static cardinalToDegree(cardinal) {
     const cardinalUpperCase = `${cardinal}`.toUpperCase();
-    if (cardinalUpperCase === 'N') {
-      return 0;
-    }
-    if (cardinalUpperCase === 'NE') {
-      return 45;
-    }
-    if (cardinalUpperCase === 'E') {
-      return 90;
-    }
-    if (cardinalUpperCase === 'SE') {
-      return 135;
-    }
-    if (cardinalUpperCase === 'S') {
-      return 180;
-    }
-    if (cardinalUpperCase === 'SW') {
-      return 225;
-    }
-    if (cardinalUpperCase === 'W') {
-      return 270;
-    }
-    if (cardinalUpperCase === 'NW') {
-      return 315;
-    }
-    if (cardinalUpperCase === 'NNE') {
-      return 22;
-    }
-    if (cardinalUpperCase === 'ENE') {
-      return 67;
-    }
-    if (cardinalUpperCase === 'ESE') {
-      return 112;
-    }
-    if (cardinalUpperCase === 'SSE') {
-      return 157;
-    }
-    if (cardinalUpperCase === 'SSW') {
-      return 202;
-    }
-    if (cardinalUpperCase === 'WSW') {
-      return 247;
-    }
-    if (cardinalUpperCase === 'WNW') {
-      return 292;
-    }
-    if (cardinalUpperCase === 'NNW') {
-      return 337;
+    const index = 'N NNE NE ENE E ESE SE SSE S SSW SW WSW W WNW NW NNW'.split(' ').indexOf(cardinalUpperCase);
+    if (index === -1) {
+      return undefined;
     }
-    return undefined;
+    const degreesTimesTwo = index * 45;
+    // integer floor
+    return degreesTimesTwo % 2 === 0 ? degreesTimesTwo / 2 : (degreesTimesTwo - 1) / 2;
```

Comment:
Code is yours, so it's up to you. My general opinion is the new code is too complex and therefore harder to maintain, and the code is slower than the original.

The other method I was considering was a JS Map:

```js
const cardinalToDegreeLookup = new Map([['N', 0], ['NE', 45], ['E', 90], ['SE', 135], ['S', 180], ['SW', 225], ['W', 270], ['NW', 315], ['NNE', 22], ['ENE', 67], ['ESE', 112], ['SSE', 157], ['SSW', 202], ['WSW', 247], ['WNW', 292], ['NNW', 337]]);

/**
 * Convert a cardinal direction to degrees.
 * North is zero and values increase clockwise.
 *
 * @param {string} cardinal - the direction.
 *
 * @return {int} degrees
 */
static cardinalToDegree(cardinal) {
    const cardinalUpperCase = `${cardinal}`.toUpperCase();
    if (cardinalToDegreeLookup.has(cardinalUpperCase)) { return cardinalToDegreeLookup.get(cardinalUpperCase); }
return undefined;
}
```

The `Map()` style is clean and understandable, but slower than using a `if/return, if/return, if/return` pattern. This function is only called when a direction is given, so speed likely isn't too important.

Reply:
Technically, the `.has()` is unneeded as you could just have `return cardinalToDegreeLookup.get(cardinalUpperCase);` as `Map.get()` returns `undefined` for missing keys.

Generally I prefer to make the `return undefined` explicit.

Reply:
I felt the in-then-else was just too bulky. I’m a math guy, not a CS guy, so my first instinct was to reduce it to a simple math problem. If you say it’s slow, I trust you. I like the map idea. If you had originally submitted that, I probably wouldn’t have messed with it.